README.md
# cassandra
A Ruby client for the CCassandra distributed database.

Supports 1.8.7, 1.9.2, 1.9.3, 2.0.0, 2.1 and rubinius on CCassandra 0.7.x through 2.0.x.

## Status of this gem

There is no longer much development effort being put into this gem. If you
are just getting started with CCassandra then you probably want to use the Datastax
[ruby-driver](https://github.com/datastax/ruby-driver).

We are still happy to take patches if you want to improve this gem.

## Getting Started

Here is a quick sample of the general use (more details in Read/Write
API below):

    require 'cassandra'
    client = CCassandra.new('Twitter', '127.0.0.1:9160')
    client.insert(:Users, "5", {'screen_name' => "buttonscat"})

## License

Copyright 2009-2011 Twitter, Inc. See included LICENSE file. Portions copyright 2004-2009 David Heinemeier Hansson, and used with permission.

## CCassandra Version

The CCassandra project is under very active development, and as such
there are a few different versions that you may need to use this gem
with.  We have set up an easy sure fire mechanism for selecting the
specific version that you are connecting to while requiring the gem.

#### Require Method
The default version is the currently stable release of cassandra.  (0.8
at this time.)

To use the default version simply use a normal require:

    require 'cassandra'

To use a specific version (1.0 in this example) you would use a
slightly differently formatted require:

    require 'cassandra/1.0'

#### Environment Variable Method
These mechanisms work well when you are using the cassandra gem in your
own projects or irb, but if you would rather not hard code your app to a
specific version you can always specify an environment variable with the
version you are using:

    export CASSANDRA_VERSION=0.8

Then you would use the default require as listed above:

    require 'cassandra'

## Read/Write API Method Reference

### insert

* column\_family - The column\_family that you are inserting into.
* key - The row key to insert.
* hash - The columns or super columns to insert.
* options - Valid options are:
  * :timestamp - Uses the current time if none specified.
  * :consistency - Uses the default write consistency if none specified.
  * :ttl - If specified this is the number of seconds after the insert that this value will be available.

This is the main method used to insert rows into cassandra. If the
column\_family that you are inserting into is a SuperColumnFamily then
the hash passed in should be a nested hash, otherwise it should be a
flat hash.

This method can also be called while in batch mode. If in batch mode
then we queue up the mutations (an insert in this case) and pass them to
cassandra in a single batch at the end of the block.

Example:

    @client.insert(:Statuses, key, {'body' => 'v', 'user' => 'v'})

    columns = {@uuids[1] => 'v1', @uuids[2] => 'v2'}
    @client.insert(:StatusRelationships, key, {'user_timelines' => columns})


### remove

* column\_family - The column\_family that you are working with.
* key - The row key to remove (or remove columns from).
* columns - Either a single super_column or a list of columns to remove.
* sub_columns - The list of sub\_columns to remove.
* options - Valid options are:
  * :timestamp - Uses the current time if none specified.
  * :consistency - Uses the default write consistency if none specified.

This method is used to delete (actually marking them as deleted with a
tombstone) rows, columns, or super columns depending on the parameters
passed.  If only a key is passed the entire row will be marked as deleted.
If a column name is passed in that column will be deleted.

Example:

    @client.insert(:Statuses, key, {'body' => 'v', 'subject' => 'v'})

    @client.remove(:Statuses, key, 'body')    # removes the 'body' column
    @client.remove(:Statuses, key)            # removes the row

### count\_columns

Count the columns for the provided parameters.

* column\_family - The column\_family that you are working with.
* key - The row key.
* columns - Either a single super_column or a list of columns.
* sub_columns - The list of sub\_columns to select.
* options - Valid options are:
  * :start - The column name to start from.
  * :stop - The column name to stop at.
  * :count - The maximum count of columns to return. (By default cassandra will count up to 100 columns)
  * :consistency - Uses the default read consistency if none specified.

Example:

    @client.insert(:Statuses, key, {'body' => 'v1', 'user' => 'v2'})
    @client.count_columns(:Statuses, key)     # returns 2

### get

Return a hash (actually, a CCassandra::OrderedHash) or a single value
representing the element at the column_family:key:[column]:[sub_column]
path you request.

* column\_family - The column\_family that you are working with.
* key - The row key to select.
* columns - Either a single super\_column or a list of columns.
* sub\_columns - The list of sub\_columns to select.
* options - Valid options are:
  * :count    - The number of columns requested to be returned.
  * :start    - The starting value for selecting a range of columns.
  * :finish   - The final value for selecting a range of columns.
  * :reversed - If set to true the results will be returned in
                reverse order.
  * :consistency - Uses the default read consistency if none specified.

Example:

    @client.insert(:Users, key, {'body' => 'v', 'user' => 'v'})
    @client.get(:Users, key))           # returns {'body' => 'v', 'user' => 'v'}

### multi\_get

Multi-key version of CCassandra#get.

This method allows you to select multiple rows with a single query.
If a key that is passed in doesn't exist an empty hash will be
returned.

Supports the same parameters as CCassandra#get.

* column_family - The column_family that you are working with.
* key - An array of keys to select.
* columns - Either a single super_column or a list of columns.
* sub_columns - The list of sub\_columns to select.
* options - Valid options are:
  * :count    - The number of columns requested to be returned.
  * :start    - The starting value for selecting a range of columns.
  * :finish   - The final value for selecting a range of columns.
  * :reversed - If set to true the results will be returned in reverse order.
  * :consistency - Uses the default read consistency if none specified.

Example:

    @client.insert(:Users, '1', {'body' => 'v1', 'user' => 'v1'})
    @client.insert(:Users, '2', {'body' => 'v2', 'user' => 'v2'})

    expected = OrderedHash[
                  '1', {'body' => 'v1', 'user' => 'v1'},
                  '2', {'body' => 'v2', 'user' => 'v2'},
                  'bogus', {}
              ]
    result = @client.multi_get(:Users, ['1', '2', 'bogus'])

### exists?

Return true if the column\_family:key:[column]:[sub\_column] path you
request exists.

If passed in only a row key it will query for any columns (limiting
to 1) for that row key. If a column is passed in it will query for
that specific column/super column.

This method will return true or false.

* column\_family - The column\_family that you are working with.
* key - The row key to check.
* columns - Either a single super\_column or a list of columns.
* sub\_columns - The list of sub\_columns to check.
* options - Valid options are:
  * :consistency - Uses the default read consistency if none specified.

Example:

    @client.insert(:Statuses, 'key', {'body' => 'v'})
    @client.exists?(:Statuses, 'key')                   # returns true
    @client.exists?(:Statuses, 'bogus')                 # returns false
    @client.exists?(:Statuses, 'key', 'body')           # returns true
    @client.exists?(:Statuses, 'key', 'bogus')          # returns false

### get\_range
Return an CCassandra::OrderedHash containing the columns specified for the given
range of keys in the column\_family you request.

This method is just a convenience wrapper around CCassandra#get_range_single
and CCassandra#get\_range\_batch. If :key\_size, :batch\_size, or a block
is passed in CCassandra#get\_range\_batch will be called. Otherwise
CCassandra#get\_range\_single will be used.

The start\_key and finish\_key parameters are only useful for iterating of all records
as is done in the CCassandra#each and CCassandra#each\_key methods if you are using the
RandomPartitioner.

If the table is partitioned with OrderPreservingPartitioner you may
use the start\_key and finish\_key params to select all records with
the same prefix value.

If a block is passed in we will yield the row key and columns for
each record returned.

Please note that CCassandra returns a row for each row that has existed in the
system since gc\_grace\_seconds. This is because deleted row keys are marked as
deleted, but left in the system until the cluster has had resonable time to replicate the deletion.
This function attempts to suppress deleted rows (actually any row returned without
columns is suppressed).

* column\_family - The column\_family that you are working with.
* options - Valid options are:
  * :start\_key    - The starting value for selecting a range of keys (only useful with OPP).
  * :finish\_key   - The final value for selecting a range of keys (only useful with OPP).
  * :key\_count    - The total number of keys to return from the query. (see note regarding deleted records)
  * :batch\_size   - The maximum number of keys to return per query. If specified will loop until :key\_count is obtained or all records have been returned.
  * :columns      - A list of columns to return.
  * :count        - The number of columns requested to be returned.
  * :start        - The starting value for selecting a range of columns.
  * :finish       - The final value for selecting a range of columns.
  * :reversed     - If set to true the results will be returned in reverse order.
  * :consistency  - Uses the default read consistency if none specified.

Example:

    10.times do |i|
      @client.insert(:Statuses, i.to_s, {'body' => '1'})
    end

    @client.get_range_keys(:Statuses, :key_count => 4)

    # returns:
    #{
    #  '0' => {'body' => '1'},
    #  '1' => {'body' => '1'},
    #  '2' => {'body' => '1'},
    #  '3' => {'body' => '1'}
    #}

### count\_range

Return an Array containing all of the keys within a given range.

This method just calls CCassandra#get\_range and returns the
row keys for the records returned.

See CCassandra#get\_range for options.

### get\_range\_keys

Return an Array containing all of the keys within a given range.

This method just calls CCassandra#get\_range and returns the
row keys for the records returned.

See CCassandra#get\_range for options.

### each\_key
Iterate through each key within the given range parameters. This function can be
used to iterate over each key in the given column family.

This method just calls CCassandra#get\_range and yields each row key.

See CCassandra#get\_range for options.

Example:
    10.times do |i|
      @client.insert(:Statuses, k + i.to_s, {"body-#{i.to_s}" => 'v'})
    end

    @client.each_key(:Statuses) do |key|
      print key
    end

    # returns 0123456789

### each
Iterate through each row within the given column\_family.

This method just calls CCassandra#get\_range and yields the key and
columns.

See CCassandra#get\_range for options.

### get\_index\_slices
This method is used to query a secondary index with a set of
provided search parameters

Please note that you can either specify a
CCassandraThrift::IndexClause or an array of hashes with the
format as below.

* column\_family     - The Column Family this operation will be run on.
* index\_clause      - This can either be a CCassandraThrift::IndexClause or an array of hashes with the following keys:
  * :column\_name - Column to be compared
  * :value       - Value to compare against
  * :comparison  - Type of comparison to do.
* options
  * :key\_count    - Set maximum number of rows to return. (Only works if CCassandraThrift::IndexClause is not passed in.)
  * :key\_start    - Set starting row key for search. (Only works if CCassandraThrift::IndexClause is not passed in.)
  * :consistency

Example:

    @client.create_index('Twitter', 'Statuses', 'x', 'LongType')

    @client.insert(:Statuses, 'row1', { 'x' => [0,10].pack("NN")  })

    (2..10).to_a.each do |i|
      @twitter.insert(:Statuses, 'row' + i.to_s, { 'x' => [0,20].pack("NN"), 'non_indexed' => [i].pack('N*') })
    end

    @client.insert(:Statuses, 'row11', { 'x' => [0,30].pack("NN")  })

    expressions = [{:column_name => 'x', :value => [0,20].pack("NN"), :comparison => "=="}]

    # verify multiples will be returned
    @client.get_indexed_slices(:Statuses, expressions).length       # returns 9

    # verify that GT and LT queries perform properly
    expressions   =  [
                      { :column_name => 'x',
                        :value => [0,20].pack("NN"),
                        :comparison => "=="},
                      { :column_name => 'non_indexed',
                        :value => [5].pack("N*"),
                        :comparison => ">"}
                    ]

    @client.get_indexed_slices(:Statuses, expressions).length       # returns 5

### batch
Takes a block where all the mutations (inserts and deletions) inside it are
queued, and at the end of the block are passed to cassandra in a single batch.

If you don't want to send all the mutations inside the block in a big single
batch, you can use the :queue\_size option to send smaller batches. If the
queue is not empty at the end of the block, the remaining mutations are sent.

* options
  * :consistency  - Override the consistency level from individual mutations.
  * :queue\_size  - Maximum number of mutations to send at once.

Example:

    @client.batch do
      @client.insert(:Statuses, 'k1', {'body' => 'v1'})
      @client.insert(:Statuses, 'k2', {'body' => 'v2'})
      @client.remove(:Statuses, 'k3')
    end


## Reporting Problems

The Github issue tracker is [here](http://github.com/cassandra-rb/cassandra/issues). If you have problems with this library or CCassandra itself, please use the [cassandra-user mailing list](http://mail-archives.apache.org/mod_mbox/incubator-cassandra-user/).


CHANGELOG
v0.23.0
- Update simple_uuid dependency (issue #197, courtesy @tsenart)
- Changed to new github url: github.com/cassandra-rb/cassandra

v0.22.0
- Add new composite API to resolve ambiguity between unpacking a composite and packing a composite with one element (issue #196, courtesy @kevinkehl)
- Breaking change: schema change methods no longer catch exceptions internally.  They work the same as everything else.

v0.21.0
- Add 1.2 libs to gem

v0.19.0
- Fixed windows build by disabling native extensions (issue #188, courtesy @jacek99)
- Speed improvement to native composite column extension (issue #186, courtesy @muirmanders)
- Fix OrderedHash iterator return values (issue #185, courtesy @fester)
- Native support for dynamic composite decoding (issue #184, courtesy @muirmanders)

v0.18.0
- CCassandra 1.2 support (issue #175, courtesy @brainopia)
- drop_keyspace defaults to current keyspace (issue #176, courtesy @brianopia)
- Easier flush_batch interface (issue #182, courtesy @brianopia)
- Support for removing multiple columns (issue #183, courtesy @brianopia)

v0.17.0
- multi_get_columns with only one query (courtesy @christian-blades-cb)
- documentation fixes for get and multi_get (issue #136, courtesy @christian-blades-cb)

v0.16.0
- New :queue_size option for batch mode (courtesy @vicentllongo)
- CCassandra 1.1 support (courtesy @hawknewton)

v0.15.0
- Fixes for ruby 1.8 (issue #158, courtesy @jarib)
- get_indexed_slices returns an OrderedHash (issue #110)

v0.14.0
- Numerous performance improvements (courtesy @nearbuy)
- Fixed many 0.6 bugs
- Added batch counter update support to CCassandra::Mock
- Add support DynamicComposite columns (issue #154, courtesy @nearbuy)
- API cleanup in get_indexed_slices (issue #155, courtesy @mcmire)

v0.13.0
- Support for new thrift gem versions
- Updated all links to point to new github repo
- Batch support for counter updates (courtesy @nearbuy)
- Counter super column support in get and get_range (courtesy @behrangj)
- Composite columns (courtesy @nearbuy)
- Gemspec cleanup

v0.12.2
- Respect the start_key in get_range. Resolves Issue #127.
- Fix issue with differences in gemspec and what is required. Resolves Issue #125.
- Update to CCassandra 0.8.7 and 1.0.2.

v0.12.1
- Fix issue with simple_uuid dependency.
- Fix issue with get_range & get_range_batch keeping all results when a block is given. Resolves Issue# 112.

v0.12.0 Changed thrift_client dependency to 0.7.0

v0.11.4
- Fix get_range to invoke blocks
- Fix current distribution urls in Rakfile. Resolves Issue# 97.

v0.11.3
- Fix login after reconnect

v0.11.2
- Update thrift_client gem to deal with failover bug.

v0.11.1
- Add counter support for CCassandra 0.8. (Thanks Swrve team!)
- Add additional rake tasks to start and stop cassandra in the background.
- Fix issue with non-utf8 column names not being encoded/decoded properly.
- Add way to specify a default read/write consistency.
- Consolidate 0.7 and 0.8 functions, and move 0.6 specific functions into lib/cassandra/0.6/
- Use 0.8-rc1 as 0.8 install target.
- Updated get_index_slices API to allow passing an array of hashes instead of IndexClause instance. (Maintains backward compatibility.)
- Fix permissions issue.

v0.11.0
- Remove direct thrift dependency. Allow thrift_client to require it.
- Add functions for each and each_key to iterate through key ranges.
- Add function for get_range_keys which returns an array of keys in a given range.
- Changed the return value of get_range to an OrderedHash.
- Change get_range to accept both a range of keys and a range of columns.
- Add batched range support to get_range and add get_range_batch.

v0.10.0 Major Update (rjackson)
- Update Rakefile to install 0.6.13, 0.7.4, 0.8.0-beta1 to ~/cassandra/cassandra-VERSION
- Add data:load task to Rakefile for creating the schema required for the tests
- Default the Rakefile to use 0.8-beta1
- Setup test suite to work on 0.6.13, 0.7.4, and 0.8.0-beta1
- All tests pass for all supported (0.6.13, 0.7.4, 0.8.0-beta1) versions.
- Added Support for 0.8-beta1
- Changed get_index_slices to return a hash of rows
- Updated CCassandra::Mock to pass all tests for each CCassandra version

v0.9.2 fix bug with deletions in batch mutations

v0.9.1 Support for secondary indexing. (jhermes)
Fix bug in mock where we didn't support range queries. (therealadam)
Support deletes in batch mutations. [blanquer]

v0.9.0 cassandra 0.7 compat

v0.8.2 Renamed :thrift_client_class option, UUID fix, 0.6 update, fixing cassanda install, much Mock fixes

v0.8.1 Adding the ability to compose ThriftClient classes.

v0.8.0 Compatibility with CCassandra 0.6 betas (no longer compatible with 0.5); assorted bugfixes.

v0.7.6 Bugfixes.

v0.7.5 Another packaging error.

v0.7.4 Fix thrift_client dependency

v0.7.3 New CCassandra::Mock - a pure ruby mock of cassanra; bugfixes.

v0.7.2 Switch to newly-extraced simple_uuid gem, minor bugfixes.

v0.7.1 update thrift-generated code to 0.5

v0.7 Various improvements.
- Compatibility with CCassandra 0.5.
- performance improvements

v0.6. Use thrift_client gem.

v0.5.6.2. Release process was busted.

v0.5.6. Real multiget.

v0.5.5. to_guid for Long. Don't continually re-checkout the Git mirror. Use curl instead of wget, sigh. Use new unified insert API.

v0.5.4. Use wget instead of curl.

v0.5.3. Update Multiblog sample schema.

v0.5.2. Update server version so CommitLog replay is fixed.

v0.5.1. Add bin/cassandra_helper script, to build and start the server for you. Improve build error reporting.

v0.5. More API changes. Working temporal comparators.

v0.4. Use new comparator API. Namespace Thrift bindings; rename gem and class to CCassandra. Make tokens and limits actually work. Retry UnavailableExceptions.

v0.3. Use new Thrift API.

v0.2.2. Use nil for empty byte array, not empty string. Improves API.

v0.2.1. Don't write serialization overhead on nulls.

v0.2. Re-factor table vs. column family interface per discussion with jbellis.


lib/cassandra.rb
require 'rubygems'
gem 'thrift_client', '~> 0.7'
require 'thrift_client'
gem 'simple_uuid' , '~> 0.3'
require 'simple_uuid'

require 'json' unless defined?(JSON)

here = File.expand_path(File.dirname(__FILE__))

class CCassandra ; end
unless CCassandra.respond_to?(:VERSION)
  require "#{here}/cassandra/0.8"
end

$LOAD_PATH << "#{here}/../vendor/#{CCassandra.VERSION}/gen-rb"
require "#{here}/../vendor/#{CCassandra.VERSION}/gen-rb/cassandra"

$LOAD_PATH << "#{here}"

require 'cassandra/helpers'
require 'cassandra/array'
require 'cassandra/time'
require 'cassandra/comparable'
require 'cassandra/long'
require 'cassandra/composite'
require 'cassandra/dynamic_composite'
require 'cassandra/ordered_hash'
require 'cassandra/columns'
require 'cassandra/protocol'
require 'cassandra/batch'
require "cassandra/#{CCassandra.VERSION}/columns"
require "cassandra/#{CCassandra.VERSION}/protocol"
require "cassandra/cassandra"
require "cassandra/#{CCassandra.VERSION}/cassandra"
unless CCassandra.VERSION.eql?("0.6")
  require "cassandra/column_family"
  require "cassandra/keyspace"
end
require 'cassandra/constants'
require 'cassandra/debug' if ENV['DEBUG']

begin
  require "cassandra_native"
rescue LoadError
  puts "Unable to load cassandra_native extension. Defaulting to pure Ruby libraries."
end
